### ISSUE TITLE
Migrate discourse.webmaker.org to use Webmaker for single sign-on

### ISSUE THREAD

**Issue by omnignorant** [mozilla/teach.webmaker.org#930]
I was trying to show folks how to create a discourse account today and ran into a snafu. the Webmaker integration is down, only google works.

@thornet @LauraHilliger @ldecoursy @toolness

**Comment by ldecoursy** [comment]
@toolness can you confirm if this is related to #931 or if this should be filed in a different repo?

**Comment by toolness** [comment]
Yeah, I have no idea what's causing this one unfortunately, as I've never touched that code. I don't even know what the proper repo to file that bug in is. :(

**Comment by LauraHilliger** [comment]
cc @secretrobotron @jbuck Not sure how long this has been down, likely since we shipped the new login? I've made sure you both have Admin rights (both your accounts are using Mofo addys). 

Lemme know if there's something I can do to help. The following URL is linked from Discourse backend and results in "Cannot GET /api/user/authenticate"

https://login.webmaker.org/api/user/authenticate

**Comment by jbuck** [comment]
We should switch webmaker discourse to use id.wm.o oauth2 login
https://meta.discourse.org/t/login-to-discourse-with-custom-oauth2-provider/14717

**Comment by ldecoursy** [comment]
@toolness or @jbuck who's the best person to help us with this?

**Comment by hannahkane** [comment]
@jbuck - this seems like it should be in a different repo. Can you advise, please? Thanks!

**Comment by jbuck** [comment]
This is going to require development time. Realistically I wouldn't be able to get to this for a month, so I'd get this in a heartbeat issue or seeing if anyone else has some spare time. In the meantime, the Webmaker signin should probably be disabled.

**Comment by toolness** [comment]
Hmm, given the fact that it'll require development time, I recommend reconsidering switching to using Discourse SSO as described in https://github.com/mozilla/teach.webmaker.org/issues/461#issuecomment-88046506. While we'd face some issues with migrating folks who have *only* logged in via Google to the existing discourse, in the long run I think it'd be a good move and would allow us to have much better integration between the teach site and the discourse in the future.

Also, if we went the Discourse SSO route, I could just implement the SSO endpoint on teach API. I've already done this for the hive directory so I'm familiar with it, and it's not really that hard.

The main challenge would really be figuring out what to do about the users who have been logging in via google auth. If they happened to use the same gmail address as their webmaker account *and* they chose the same username on discourse as their webmaker account, then we're fine, but if they used a different email and/or a different username, then, um, we'd have to manually migrate them over on a case-by-case basis or something.

**Comment by hannahkane** [comment]
I'm okay with that plan, but would like to hash out the migration step further. @toolness, can you look into it and clarify further what the options might be for migration, especially in terms of the end user's experience?

**Comment by toolness** [comment]
It's kind of messy, I guess--I think the first step would be running an analysis of our current userbase on discourse, essentially looking at anyone who's ever participated in any way, and mapping their username+email to the existing webmaker user table. Anything that doesn't match likely means that the user has been logging in via google.

If that list is really small, then we can probably just manually contact those individuals and migrate them manually, e.g. by having them tell us their webmaker username, and then us changing their username/email in Discourse to match the webmaker info.

If the list is big, then things get more complicated. We could have some sort of filtering criteria where e.g. if the user hasn't posted to our discourse in the past year, we just delete their discourse account. If the filtered list is small then we can proceed with the manual process I described above.

If the list isn't small and there's no good filtering criteria, then I'm not really sure what to do that wouldn't be really weird or inconvenient in some way. For instance, we could rename discourse google login users to "google_*username*", create accounts for each of them on webmaker, and then email them.

**Comment by toolness** [comment]
We could also mix-and-match these options: for example, for anyone we manually contact that doesn't respond within a given time frame, we could then prefix their discourse name with "google_" and create a webmaker login for them.

**Comment by hannahkane** [comment]
Let's start with analyzing the list, and seeing what we're working with then.

**Comment by toolness** [comment]
Ok, I've got something!

I wrote this [`process-discourse-users.js`](https://gist.github.com/toolness/a1d2ac8c1400302b36ca) script that generated the following report:

----

1. Discourse users with identical Webmaker usernames: 357
2. Discourse users with reconcilable Webmaker usernames: 530
3. Discourse users with conflicting Webmaker usernames: 4
  * [iamrowdymehul] / [rowdymehul]
  * [aini] / [nizibaba]
  * [kmiklasz] / [kamman13]
  * [amirad] / [amira]
4. Discourse users without a known webmaker account: 130
  * Who have posted at least once: 48
    * [Yaqila_Group]
    * [Kasandra_Posey]
    * [faberglas]
    * [John_Mongan]
    * [khrbssmor]
    * [adaminfinitum]
    * [Jeremiah_Pearson]
    * [RaziKallayi]
    * [Errin_Johnson]
    * [Bhushan3d]
    * [michael_delgado]
    * [Amit_123]
    * [mrnavas]
    * [Sra_Silva]
    * [marta_kebre]
    * [WiFi_Club]
    * [Keri]
    * [deepak_c_blr]
    * [wor_tony]
    * [rethinktheboxdesign]
    * [jeremiak]
    * [jen061191]
    * [Prasoon]
    * [Amit]
    * [smadisonboyd]
    * [MrVinitBansal]
    * [matiasj23]
    * [sillymind69]
    * [JGSokomato]
    * [chintu_nitin]
    * [alvarado_javier]
    * [mclato]
    * [stevenkinney]
    * [laia]
    * [Eduitynet]
    * [lucid_elmasry]
    * [alexlearning]
    * [joanvc]
    * [cheikhomar]
    * [irasocol]
    * [mikko_kontto]
    * [mr_cool3_tony]
    * [emgollie]
    * [Robelia]
    * [lynncasper]
    * [evanjones]
    * [Mozzadrella]
    * [Mermi]

[iamrowdymehul]: https://discourse.webmaker.org/users/iamrowdymehul
[rowdymehul]: https://discourse.webmaker.org/users/rowdymehul
[aini]: https://discourse.webmaker.org/users/aini
[nizibaba]: https://discourse.webmaker.org/users/nizibaba
[kmiklasz]: https://discourse.webmaker.org/users/kmiklasz
[kamman13]: https://discourse.webmaker.org/users/kamman13
[amirad]: https://discourse.webmaker.org/users/amirad
[amira]: https://discourse.webmaker.org/users/amira
[Yaqila_Group]: https://discourse.webmaker.org/users/Yaqila_Group
[Kasandra_Posey]: https://discourse.webmaker.org/users/Kasandra_Posey
[faberglas]: https://discourse.webmaker.org/users/faberglas
[John_Mongan]: https://discourse.webmaker.org/users/John_Mongan
[khrbssmor]: https://discourse.webmaker.org/users/khrbssmor
[adaminfinitum]: https://discourse.webmaker.org/users/adaminfinitum
[Jeremiah_Pearson]: https://discourse.webmaker.org/users/Jeremiah_Pearson
[RaziKallayi]: https://discourse.webmaker.org/users/RaziKallayi
[Errin_Johnson]: https://discourse.webmaker.org/users/Errin_Johnson
[Bhushan3d]: https://discourse.webmaker.org/users/Bhushan3d
[michael_delgado]: https://discourse.webmaker.org/users/michael_delgado
[Amit_123]: https://discourse.webmaker.org/users/Amit_123
[mrnavas]: https://discourse.webmaker.org/users/mrnavas
[Sra_Silva]: https://discourse.webmaker.org/users/Sra_Silva
[marta_kebre]: https://discourse.webmaker.org/users/marta_kebre
[WiFi_Club]: https://discourse.webmaker.org/users/WiFi_Club
[Keri]: https://discourse.webmaker.org/users/Keri
[deepak_c_blr]: https://discourse.webmaker.org/users/deepak_c_blr
[wor_tony]: https://discourse.webmaker.org/users/wor_tony
[rethinktheboxdesign]: https://discourse.webmaker.org/users/rethinktheboxdesign
[jeremiak]: https://discourse.webmaker.org/users/jeremiak
[jen061191]: https://discourse.webmaker.org/users/jen061191
[Prasoon]: https://discourse.webmaker.org/users/Prasoon
[Amit]: https://discourse.webmaker.org/users/Amit
[smadisonboyd]: https://discourse.webmaker.org/users/smadisonboyd
[MrVinitBansal]: https://discourse.webmaker.org/users/MrVinitBansal
[matiasj23]: https://discourse.webmaker.org/users/matiasj23
[sillymind69]: https://discourse.webmaker.org/users/sillymind69
[JGSokomato]: https://discourse.webmaker.org/users/JGSokomato
[chintu_nitin]: https://discourse.webmaker.org/users/chintu_nitin
[alvarado_javier]: https://discourse.webmaker.org/users/alvarado_javier
[mclato]: https://discourse.webmaker.org/users/mclato
[stevenkinney]: https://discourse.webmaker.org/users/stevenkinney
[laia]: https://discourse.webmaker.org/users/laia
[Eduitynet]: https://discourse.webmaker.org/users/Eduitynet
[lucid_elmasry]: https://discourse.webmaker.org/users/lucid_elmasry
[alexlearning]: https://discourse.webmaker.org/users/alexlearning
[joanvc]: https://discourse.webmaker.org/users/joanvc
[cheikhomar]: https://discourse.webmaker.org/users/cheikhomar
[irasocol]: https://discourse.webmaker.org/users/irasocol
[mikko_kontto]: https://discourse.webmaker.org/users/mikko_kontto
[mr_cool3_tony]: https://discourse.webmaker.org/users/mr_cool3_tony
[emgollie]: https://discourse.webmaker.org/users/emgollie
[Robelia]: https://discourse.webmaker.org/users/Robelia
[lynncasper]: https://discourse.webmaker.org/users/lynncasper
[evanjones]: https://discourse.webmaker.org/users/evanjones
[Mozzadrella]: https://discourse.webmaker.org/users/Mozzadrella
[Mermi]: https://discourse.webmaker.org/users/Mermi

----

**Group 1** consists of people who have Webmaker and Discourse accounts with the exact same username and email. These are the ones that are effortless to migrate.

**Group 2** consists of people who have Webmaker and Discourse accounts with the exact same email, but differing usernames. These can be relatively easily migrated because we can just rename their Discourse username to their Webmaker username.

**Group 3** consists of people who have Webmaker and Discourse accounts with the exact same email, but differing usernames, like Group 2. However, unlike Group 2, their Discourse accounts can't simply be renamed because the username is taken. It's quite likely that these are just people who have logged in via *both* Webmaker and Google in the past, though (it's almost certainly the case with Amira, who is in this group). Since there's only a handful of users in this group, it might be easiest to just contact them individually. We could also see if any of the conflicting usernames have never posted anything, and if so, we can just delete them outright.

**Group 4** consists of people who don't have a Webmaker account with the same email address as their Discourse account. It's possible that they might be present on both Webmaker and Discourse, but via different emails; alternatively, it's possible that they never used Webmaker but participated in Discourse conversations. The ones in this group who have made at least one post are the ones that we really want to investigate--we're probably OK just deleting the rest.

**Comment by hannahkane** [comment]
Thanks, @toolness. This doesn't seem that bad. 

@amirad, @ldecoursy, @thornet  - how do you want to handle groups 3 and 4? We could send PMs to everyone in those groups? (I'd probably vote for just the 48 in group 4 who've posted in the past.)

**Comment by amirad** [comment]
Man, who is this amirad/amira and why is her user account giving us such a problem ;) Okay yes, it's probably because people like me have definitely logged in with both Webmaker and Google. 

Group 1 + 2, I would just migrate. 
Group 3 I would contact those people and let them know. 
Group 4 I would email only those 48 people who has posted before.

**Comment by ldecoursy** [comment]
+1, although wonder if we should also alert people in group 2 that we're changing their discourse username...

**Comment by toolness** [comment]
Er wait, now I'm realizing something... [according to the co-founder of Discourse](https://meta.discourse.org/t/ability-to-merge-users/9220/3):

> The email address is your identity from our perspective.

That implies that there is only one user per email address in Discourse, yet [we don't have such a restriction in Webmaker](https://github.com/mozilla/id.webmaker.org/issues/120#issuecomment-88969419). This could make it really hard to offer Webmaker-based login, unless we have some sort of concept of a "primary" account for an email address, or use some other sort of workaround. 

Indeed, it seems a broader solution might be necessary for integrating with lots of third-party services and/or software--[I've already found a similar problem with Cred.ly integration](https://github.com/mozilla/teach.webmaker.org/issues/710#issuecomment-99420693).

**Comment by jbuck** [comment]
@toolness We ended up not implementing that feature, right now, emails are unique, you can't duplicate them.

**Comment by toolness** [comment]
But won't the emails *eventually* not be unique? I don't want to implement something that breaks a few months down the line...

Or is it possible for us to add some kind of "must still work with discourse/credly/etc" requirement to the "support multiple accounts per email" feature that might eventually get implemented? If we can either be assured that such a feature won't be implemented in the forseeable future and/or that a workaround will be implemented on id.webmaker.org if such a feature should ever exist (e.g. as part of the OAuth flow), that would be really great...

**Comment by toolness** [comment]
Hey @thisandagain, any idea if the "multiple users per email" feature will end up being implemented? And/or if there will be some kind of "primary username" concept that maps an email address to one username?

A number of third-party services/packages we'd like to integrate with (Discourse and Cred.ly, mentioned above) have a 1:1 mapping between usernames and emails that is incompatible with the "multiple users per email" model, and I'm not sure how to resolve it.

**Comment by thisandagain** [comment]
I got quite a bit of push-back on the "multiple users per email" feature and so I don't have it on the roadmap any longer. We are ok to move forward with a 1:1 mapping. Thanks for the follow-up!

**Comment by hannahkane** [comment]
@toolness - so, no longer blocked on this one? Let me know what you need from me to proceed.

**Comment by toolness** [comment]
Ah thanks @thisandagain--bummer that you got pushback, since it does seem like a very useful feature, but at least it will make 3rd-party integration easier... 

(BTW you can unsubscribe from this issue now if you don't want the spam, I think that's all we needed from you.)

@hannahkane, I *think* I'm no longer blocked on this one. The only additional thing to note is that id.webmaker.org doesn't currently require users to validate their email address in any way, so I want to make sure it's not possible for people to somehow "impersonate" existing discourse users via webmaker login. But I think that's about it...

**Comment by toolness** [comment]
Ok, I did a bit more digging on the Webmaker side and found out that most of the 48 people in group 4--the ones who have no corresponding Webmaker account--actually have their username available on the Webmaker side. For instance, "stevenkinney" doesn't have a Webmaker account with the same email, but the username "stevenkinney" is available in Webmaker, so we could potentially create a new Webmaker user with the same username and email as his Discourse account.

If we did that, then we're only left with a tiny handful of users who have their Discourse usernames taken in Webmaker by accounts that are associated with different emails:

      * [faberglas]
      * [mrnavas]
      * [Prasoon]
      * [Amit]
      * [mclato]
      * [emgollie]
      * [lynncasper]
      * [Mozzadrella]
      * [Mermi]

A number of these are folks I know; after looking at the email addresses for both accounts, it's clear that emgollie, lynncasper, and Mozzadrella created Webmaker accounts with their work email addresses, yet logged into Discourse with their non-work Google accounts. It's likely that the others in the list have done the same. In any case, if we want to move as fast as possible and block on as few people as possible, we could just limit ourselves to only emailing the above list of nine people, and the four people in group 3. For the rest, we'd automatically do one of:

* Just migrating them over transparently b/c they have identical Webmaker+Discourse usernames with the same emails (Group 1)
* Renaming their Discourse username to match their Webmaker username (Group 2) and emailing them about the change (as lainie [suggested](https://github.com/mozilla/teach.webmaker.org/issues/930#issuecomment-112072990))
* Creating Webmaker accounts to match their Discourse accounts (most of Group 4) and emailing them about the change, giving them a link to set their password

**Comment by toolness** [comment]
Ok, slightly changed plan... here's the latest output of my script for reference:

----






1. Discourse users with identical Webmaker usernames: 357
2. Discourse users with reconcilable Webmaker usernames: 530
3. Discourse users with conflicting Webmaker usernames: 4
  * [iamrowdymehul] / [rowdymehul]
  * [aini] / [nizibaba]
  * [kmiklasz] / [kamman13]
  * [amirad] / [amira]
4. Discourse users without a known webmaker account: 130
  * Who can't have new Webmaker accounts created for them: 
    * [Zao777] - never posted
    * [CRUSADERED] - never posted
    * [111] - never posted
    * [rowdymehul] - never posted
    * [martin] - never posted
    * [Rudloff] - never posted
    * [susan] - never posted
    * [1l2p] - never posted
    * [amira] - never posted
    * [Mozzadrella] - last posted Apr 2014
    * [emgollie] - last posted Jun 2014
    * [mclato] - last posted Aug 2014
    * [Mermi] - last posted Aug 2014
    * [Amit] - last posted Aug 2014
    * [Prasoon] - last posted Sep 2014
    * [lynncasper] - last posted Oct 2014
    * [mrnavas] - last posted Apr 2015
    * [faberglas] - last posted May 2015

[111]: https://discourse.webmaker.org/users/111
[iamrowdymehul]: https://discourse.webmaker.org/users/iamrowdymehul
[rowdymehul]: https://discourse.webmaker.org/users/rowdymehul
[aini]: https://discourse.webmaker.org/users/aini
[nizibaba]: https://discourse.webmaker.org/users/nizibaba
[kmiklasz]: https://discourse.webmaker.org/users/kmiklasz
[kamman13]: https://discourse.webmaker.org/users/kamman13
[amirad]: https://discourse.webmaker.org/users/amirad
[amira]: https://discourse.webmaker.org/users/amira
[Zao777]: https://discourse.webmaker.org/users/Zao777
[CRUSADERED]: https://discourse.webmaker.org/users/CRUSADERED
[martin]: https://discourse.webmaker.org/users/martin
[Rudloff]: https://discourse.webmaker.org/users/Rudloff
[susan]: https://discourse.webmaker.org/users/susan
[1l2p]: https://discourse.webmaker.org/users/1l2p
[Mozzadrella]: https://discourse.webmaker.org/users/Mozzadrella
[emgollie]: https://discourse.webmaker.org/users/emgollie
[mclato]: https://discourse.webmaker.org/users/mclato
[Mermi]: https://discourse.webmaker.org/users/Mermi
[Amit]: https://discourse.webmaker.org/users/Amit
[Prasoon]: https://discourse.webmaker.org/users/Prasoon
[lynncasper]: https://discourse.webmaker.org/users/lynncasper
[mrnavas]: https://discourse.webmaker.org/users/mrnavas
[faberglas]: https://discourse.webmaker.org/users/faberglas




----

We won't be deleting any Discourse users, at least not in an automated way. That relieves me of having to write code that automates the process, and being terrified of it accidentally mis-firing and mass deleting the wrong users. :sweat: 

That leaves us with the following types of *automated* user actions, which are applicable to every user not explicitly named in the above report:

* Just migrating them over transparently b/c they have identical Webmaker+Discourse usernames with the same emails (Group 1)
* Renaming their Discourse username to match their Webmaker username (Group 2) and emailing them about the change (as lainie [suggested](https://github.com/mozilla/teach.webmaker.org/issues/930#issuecomment-112072990))
* Creating Webmaker accounts to match their Discourse accounts (most of Group 4) and emailing them about the change, giving them a link to set their password

And for the users explicitly named in the above report, we'll have to manually figure something out.

**Comment by amirad** [comment]
Types of emails to send (as discussed with @toolness and @hannahkane on 06/17/15):

•you're discourse username got changed to match your webmaker username (group 1 and 2)
•we need to change your username, which username/email associated with discourse do you want to use (group 3)
•we're migrating to using Webmaker as a login, so we created an account for you there with the same account name. If you want to keep your account set a password now (group 4)
•we're migrating to using Webmaker as a login, you might already have an account on Webmaker and if you do let us know what it is so we can link it for you. If you don't respond, we're going to rename your username to something that is available (group 4b)

**Comment by toolness** [comment]
sounds good!

Hannah just mentioned that we should also tell users that this will only affect their MOZILLA LEARNING DISCOURSE login, since they might be confused about whether this change will affect any other Discourse instances they might be participating in (e.g. the [mozilla community discourse](https://discourse.mozilla-community.org/) or the open science, open news discourses, etc).

**Comment by hannahkane** [comment]
Drafted emails here: https://etherpad.mozilla.org/discourseemails

I'm sure I got something wrong, though. @toolness, can you do a sanity check, please? :)

**Comment by toolness** [comment]
Awsm!!!

Note that when I send out the automated emails, I'll only be able to easily send them in plain-text form, so could you modify them to only include URLs in the main text, i.e. not assume we can hyperlink text in the email body?

**Comment by hannahkane** [comment]
K. Modified. Still needs a sanity check! :)

**Comment by toolness** [comment]
Oh the "Awsm" was in regards to the sanity check, it looks good :)

By the way... since we're not actually validating email addresses on id.webmaker.org, we're going to have to tell Discourse to validate the user's email for us ... this is a bummer but it's the only way to do it :(

**Comment by toolness** [comment]
Ok, I added support for [Discourse SSO](https://meta.discourse.org/t/official-single-sign-on-for-discourse/13045) to teach-api in https://github.com/mozilla/teach-api/pull/18, pushed it to production, and hooked up discourse.webmaker.org to it.

Everything seems to be working.

If you're already logged-in to teach.mozilla.org, just clicking "Login" on Discourse *should* automatically redirect you back to Discourse, and show you as logged in to the same account.

If you're *not* already logged in to teach.m.o, logging in to Discourse should send you to id.webmaker.org. Logging in will then log you into *both* teach.m.o *and* Discourse.

Now we just need to do the actual [user migration](https://github.com/mozilla/teach.webmaker.org/issues/930#issuecomment-112818883).

**Comment by hannahkane** [comment]
Just tested, and flows worked as expected. Note: logging out of teach.m.o does NOT log a user out of Discourse.

**Comment by toolness** [comment]
Ah, yeah, I didn't say anything about logging out because it's so complicated lol.  :)

Discourse basically treats Webmaker login the same way it treats Google login--so logging in requires one to sign into the "identity provider" (Webmaker or Google), but once logged in, you're largely independent of your sign-in state on your identity provider.

**Comment by toolness** [comment]
To really make sure this works for brand-new users, one of us should probably create a brand-new account on Webmaker, that doesn't currently exist at all in Discourse, and try logging in and making a test post. I *think* that Discourse should force them to validate their email, but I'm not sure.

**Comment by hannahkane** [comment]
Understood. Thanks!

**Comment by toolness** [comment]
Ok, a few more notes:

* There actually *can't* be a purely 1:1 mapping between Webmaker usernames and Discourse usernames, because the username requirements are different: Discourse basically has the [same rules as Twitter](https://meta.discourse.org/t/what-are-the-rules-for-usernames/13458), while Webmaker, for instance, allows `-` but disallows all other non-alphanumeric characters.
* This doesn't mean that there can't be a 1:1 mapping between Webmaker *users* and Discourse *users*--just that the usernames might be different, but the email addresses are the same. For instance, on Discourse, the account associated with my work email is "Atul", while on Webmaker it's "atul-moz" (which can't exist on Discourse). Logging on through the new Discourse SSO endpoint works, however, because the emails are the same. In light of this, **I recommend *not* actually renaming the users in group 2**, as it's actually potentially impossible for some users; and anyways, those users in group 2 will still be able to log in, they'll just have slightly different usernames than their Webmaker one.
* I passed [`require_activation=true`](https://meta.discourse.org/t/official-single-sign-on-for-discourse/13045) to Discourse from the SSO endpoint, but it doesn't actually seem to do anything... I created a brand-new user on Webmaker and logged into Discourse with it, and I was able to make a post without needing to verify my email address. Hopefully this will not make anything explode.

**Comment by toolness** [comment]
Ok, I've created all the webmaker users in group 4, just need to email them now.

**Comment by hannahkane** [comment]
Sweet. I can do it, but how do I get their email addresses?

**Comment by toolness** [comment]
No probs, I already sent them!

Now all that's needed is to manually contact everyone whose username is mentioned in the report. I'll send you and @amirad an email with their emails.

**Comment by toolness** [comment]
Ok, I sent the email to you folks, and you're welcome to make a separate issue for that if you want to track progress on it, but since everything else is done, I'm closing this issue! woot!

**Comment by hannahkane** [comment]
woot indeed. thanks, @toolness!
